x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharTestScript.java

```diff
@@ -0,0 +1,246 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
+
+import com.carrotsearch.randomizedtesting.generators.RandomNumbers;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.util.set.Sets;
+
+import java.math.BigDecimal;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.util.Arrays.asList;
+import static java.util.regex.Pattern.quote;
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateTimeTestUtils.dateTime;
+
+/**
+ * Generates an psql file that can be used to generate the test dataset for the {@link DateTimeToCharProcessorTests}.
+ */
+public class ToCharTestScript {
+
+    private static class TestRecord {
+        private final BigDecimal secondsAndFractionsSinceEpoch;
+        private final String formatString;
+        private final String zoneId;
+
+        TestRecord(BigDecimal secondsAndFractionsSinceEpoch, String formatString) {
+            this.secondsAndFractionsSinceEpoch = secondsAndFractionsSinceEpoch;
+            this.formatString = formatString;
+            this.zoneId = ToCharTestScript.randomFromCollection(TIMEZONES_TO_TEST);
+        }
+    }
+
+    private static final long SECONDS_IN_YEAR = 365 * 24 * 60 * 60L;
+    public static final String DELIMITER = "|";
+    public static final String PATTERN_DELIMITER = " @ ";
+    // these patterns are hard to sync up between PostgreSQL and Elasticsearch, so we just warn, but actually
+    // accept the output of Elasticsearch as is
+    public static final Set<String> NOT_FULLY_MATCHABLE_PATTERNS = Set.of("TZ", "tz");
+    private static final Set<String> UNSUPPORTED_PATTERN_MODIFIERS = Set.of("FX", "TM", "SP");
+    private static final List<String> PATTERNS = new ArrayList<>(ToCharFormatter.FORMATTER_MAP.keySet());
+    private static final List<String> FILL_MODIFIERS = asList("FM", "fm", "");
+    private static final List<String> ORDINAL_SUFFIX_MODIFIERS = asList("TH", "th", "");
+    // timezones that are valid both in Java and in Postgres
+    public static final List<String> TIMEZONES_TO_TEST =
+            readAllLinesWithoutComment(ToCharTestScript.class.getResource("tochar.test_timezones.txt"));
+
+    private final List<TestRecord> testRecords = new ArrayList<>();
+    private final List<BigDecimal> testEpochSeconds = new ArrayList<>();
+
+    @SuppressForbidden(reason = "It is ok to use ThreadLocalRandom outside of an actual test")
```

Comment:
This issue is still in place - why use a `ThreadLocalRandom` when no concurrency is needed and not rely on `Random` itself?
Also try to move away from it being static as that's tricky to clean-up. The new Random can be created within the `main` method and then passed around.